<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Background" data-toc-modified-id="Background-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Background</a></span></li><li><span><a href="#Introduction" data-toc-modified-id="Introduction-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Introduction</a></span></li><li><span><a href="#Libraries" data-toc-modified-id="Libraries-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Libraries</a></span></li><li><span><a href="#Data" data-toc-modified-id="Data-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Data</a></span><ul class="toc-item"><li><span><a href="#Overview" data-toc-modified-id="Overview-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Overview</a></span></li><li><span><a href="#Missing-values" data-toc-modified-id="Missing-values-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Missing values</a></span></li><li><span><a href="#Duplicates" data-toc-modified-id="Duplicates-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Duplicates</a></span></li></ul></li><li><span><a href="#Exploratory-Data-Analysis" data-toc-modified-id="Exploratory-Data-Analysis-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Exploratory Data Analysis</a></span><ul class="toc-item"><li><span><a href="#Overview" data-toc-modified-id="Overview-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Overview</a></span></li><li><span><a href="#Feature-Engineering" data-toc-modified-id="Feature-Engineering-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Feature Engineering</a></span></li></ul></li><li><span><a href="#Modeling" data-toc-modified-id="Modeling-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Modeling</a></span><ul class="toc-item"><li><span><a href="#Data-Prepocessing" data-toc-modified-id="Data-Prepocessing-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Data Prepocessing</a></span></li><li><span><a href="#Machine-Learning" data-toc-modified-id="Machine-Learning-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Machine Learning</a></span><ul class="toc-item"><li><span><a href="#Logistic-Regression" data-toc-modified-id="Logistic-Regression-6.2.1"><span class="toc-item-num">6.2.1&nbsp;&nbsp;</span>Logistic Regression</a></span></li><li><span><a href="#Support-Vector-Machines" data-toc-modified-id="Support-Vector-Machines-6.2.2"><span class="toc-item-num">6.2.2&nbsp;&nbsp;</span>Support Vector Machines</a></span></li><li><span><a href="#KNearest-Neighbors" data-toc-modified-id="KNearest-Neighbors-6.2.3"><span class="toc-item-num">6.2.3&nbsp;&nbsp;</span>KNearest Neighbors</a></span></li><li><span><a href="#Naive-Bayes" data-toc-modified-id="Naive-Bayes-6.2.4"><span class="toc-item-num">6.2.4&nbsp;&nbsp;</span>Naive Bayes</a></span></li><li><span><a href="#Random-Forest" data-toc-modified-id="Random-Forest-6.2.5"><span class="toc-item-num">6.2.5&nbsp;&nbsp;</span>Random Forest</a></span></li></ul></li><li><span><a href="#Deep-Learning" data-toc-modified-id="Deep-Learning-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Deep Learning</a></span></li></ul></li><li><span><a href="#Results" data-toc-modified-id="Results-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Results</a></span></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Conclusion</a></span></li></ul></div>

<a href="https://colab.research.google.com/github/Maredipinkydimakatso/jhb-ss3-classification-Repo/blob/master/Classification_submission_file.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Background

* General information about climate change

# Introduction

* Information about this classification sprint 

# Libraries

In [1]:
# standard libraries
import pandas as pd
import numpy as np

# plotting libraries
import matplotlib.pyplot as plt
from matplotlib.pyplot import rcParams
import seaborn as sns
%matplotlib inline
rcParams['figure.figsize'] = 10,8
sns.set(style='whitegrid', palette='muted')

# text analysis libraries
import re
import spacy
import string

# model building and valuation libraries
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.metrics import f1_score, recall_score, precision_score, make_scorer
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfVectorizer

# other libraries
import warnings
warnings.filterwarnings('ignore')

# Data

## Overview

In [4]:
df = pd.read_csv('train.csv')

In [5]:
# Check the first five rows of the dataframe
df.head()

sentiment                                            message  tweetid
0          1  PolySciMajor EPA chief doesn't think carbon di...   625221
1          1  It's not like we lack evidence of anthropogeni...   126103
2          2  RT @RawStory: Researchers say we have three ye...   698562
3          1  #TodayinMaker# WIRED : 2016 was a pivotal year...   573736
4          1  RT @SoyNovioDeTodas: It's 2016, and a racist, ...   466954

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15819 entries, 0 to 15818
Data columns (total 3 columns):
sentiment    15819 non-null int64
message      15819 non-null object
tweetid      15819 non-null int64
dtypes: int64(2), object(1)
memory usage: 370.8+ KB


The data frame consists of three columns, sentiment which is an interger value refering to the tweet class.

| **Class** | **Description** |
| :----------: | :------------- |
| 2 | **News:** the tweet links to factual news about climate change |
| 1 | **Pro:** the tweet supports the belief of man-made climate change |
| 0 | **Neutral:** the tweet neither supports nor refutes the belief of man-made climate change |
| -1 | **Anti:** the tweet does not believe in man-made climate change |

source: https://www.kaggle.com/c/climate-change-belief-analysis/data

The message column is the tweet collected from April 2015 to February 2018 expressing the belief 
about man-made climate change. The tweetid is a unique integer value corresponding to the tweet message.

In [7]:
print(f'The dataframe consists of {len(df)} twitter messages.')

The dataframe consists of 15819 twitter messages.


## Missing values

In [8]:
# Check for NaN entries
df.isnull().sum()

sentiment    0
message      0
tweetid      0
dtype: int64

There are currently no missing values in the dataset.

In [9]:
# Sometimes empty strings can be regarded as not being null, therefore there has to be a check for empty strings message column.
blanks = []
for index, sentiment, message, tweetid in df.itertuples():
    if message.isspace():
        blanks.append(index)

In [10]:
blanks

[]

Similarly, there appears to be no empty strings in the dataset.

## Duplicates

In [11]:
all_duplicates = df[df.duplicated(['message'])]

In [13]:
all_duplicates.head(10)

sentiment                                            message  tweetid
51           1  RT @StephenSchlegel: she's thinking about how ...   785499
70           1  RT @kelkulus: Irony: Florida, a state in dange...   857540
98           1  RT @StephenSchlegel: she's thinking about how ...   603318
112          1  RT @SenSanders: We have a president-elect who ...   664093
122          1  RT @StephenSchlegel: she's thinking about how ...    97375
123          1  RT @SethMacFarlane: HRC proposes installing ha...   383885
142          1  RT @StephenSchlegel: she's thinking about how ...   188844
184          1  RT @StephenSchlegel: she's thinking about how ...   990837
224          1  RT @StephenSchlegel: she's thinking about how ...    80936
240          1  RT @StephenSchlegel: she's thinking about how ...   385351

In [20]:
len(all_duplicates)

1590

There are 1590 messages which are duplicated amongst each other in terms of the message column only.

In [15]:
duplicates = df[df.duplicated(['sentiment','message'])]

In [16]:
len(duplicates)

1586

There are 1586 messages which are duplicated both in sentiment and message.

In [22]:
print(f"Therefore {len(all_duplicates) - len(duplicates)} of the rows are duplicate messages but not sharing the\
 same sentiment, therefore they will all be dropped.")

Therefore 4 of the rows are duplicate messages but not sharing the same sentiment, therefore they will all be dropped.


In [ ]:
not_matching = [index for index in full_dupli.index if index not in dupli.index]

In [ ]:
removal_list = []
for index in not_matching:
    remove = df[df.loc[:,'message'] == df.loc[index, 'message']].index.to_list()
    removal_list += remove

In [ ]:
removal_list

In [ ]:
df.loc[removal_list, :]

# Exploratory Data Analysis

## Overview

In [ ]:
# Percentage split of the data
plt.figure(figsize=(6,6))
sns.set(font_scale=1.1)
df['sentiment'].value_counts().plot(kind='pie', autopct='%.1f', labels=['Pro','News','Neutral','Anti']);

Based on the pie chart, 50.8% of the tweets supports the belief of man-made climate change **(Pro)**, 24.9% are based on factual news about climate change **(News)**, 15.8% of the tweets are rather neutral on the subject **(Neutral)**, and 8.6% do not believe in man-made climate change **(Anti)**. 

## Feature Engineering

* EDA of new features

In [ ]:
# Loading a spacy English dictionary for NLP
nlp = spacy.load('en_core_web_md')

# Modeling

## Data Prepocessing

## Machine Learning

### Logistic Regression

### Support Vector Machines

### KNearest Neighbors

### Naive Bayes

### Random Forest

## Deep Learning

# Results

# Conclusion